python/lsst/daf/butler/_butler.py

```diff
@@ -1924,6 +1927,33 @@ def transfer_from(self, source_butler: Butler, source_refs: Iterable[DatasetRef]
         if not newly_registered_dataset_types:
             log.log(VERBOSE, "All required dataset types are known to the target Butler")
 
+        # As for dataset types, it is best if missing run collections in
+        # the target butler are created outside of a transaction.
+        # If the transfer fails this would leave empty collections behind
+        # but is safer.
+        # Also as for dataset types there is a possibility of a race
+        # with another transfer job creating the run.
+        # Reporting that a run was missing is still informative even
+        # if some other process created it in the race.
+        missing_run_collections: Set[str] = set()
+        for run in source_run_collections:
+            target_type: Optional[CollectionType]
+            try:
+                target_type = self.registry.getCollectionType(run)
+            except MissingCollectionError:
+                target_type = None
+            if target_type is None:
+                run_doc = source_butler.registry.getCollectionDocumentation(run)
+                self.registry.registerRun(run, doc=run_doc)
+                missing_run_collections.add(run)
+            elif target_type != CollectionType.RUN:
+                raise ValueError(f"Run collection {run} in source butler is not a RUN in target Butler,"
+                                 f" is {target_type.name}")
```

Comment:
Moving `DatasetType` registration out of the transaction is _probably_ good even after DM-31956 lands (imminent; just want to write a detailed reply to Andy's comments and then merge the code as-is), because it could involve creating a new table and in some databases (but not SQLite or PostgreSQL!) that's illegal inside a transaction.  But I think moving the collection registration outside the transaction is going to be totally obsoleted by DM-31956:

- There's no scenario where we shouldn't register the collection if it doesn't exist.
- The DM-31956 fix alone is not going to be vulnerable to these kinds of race conditions, and after it merges, the logic here will just open us back up to races.

I think I'd just drop this commit.